File: src/1101-The-Earliest-Moment-When-Everyone-Become-Friends/1101.js
var earliestAcq = function(logs, N) {
    logs.sort(function(a, b){
        return a[0] < b[0];
    });

    var s = [];
    for (var i = 0; i <= N; ++i) s[i] = i;
    function find(x) {
        if (s[x] != x) s[x] = find(s[x]);
        return s[x];
    };

    var cnt = 0;
    for (var i = 0; i < logs.length; ++i) {
        var rx = find(logs[i][1]), ry = find(logs[i][2]);
        if (rx != ry) {
            s[rx] = ry;
            cnt++;
        }
        if (cnt == N - 1) return logs[i][0];
    }
    return -1;
};

File: src/1101-The-Earliest-Moment-When-Everyone-Become-Friends/1101.cpp
class Solution 
{
public:
    int earliestAcq(vector<vector<int>>& logs, int N) 
    {
        sort(logs.begin(), logs.end(), [](const vector<int>& a, const vector<int>& b){
            return a[0] < b[0];
        });

        s = vector<int>(N + 1, 0);
        for (int i = 0; i <= N; ++i) s[i] = i;
        int cnt = 0;
        for (auto& log : logs)
        {
            int rx = find(log[1]), ry = find(log[2]);
            if (rx != ry)
            {
                s[rx] = ry;
                cnt++;
            }
            if (cnt == N - 1) return log[0];
        }
        return -1;
    }

    vector<int> s;
    int find(int x)
    {
        if (s[x] != x) s[x] = find(s[x]);
        return s[x];
    }
}